sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/TableHeaderHasIdOrScopeCheck.java

```diff
@@ -17,22 +17,100 @@
  */
 package org.sonar.plugins.html.checks.sonar;
 
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.List;
 import org.sonar.check.Rule;
 import org.sonar.plugins.html.checks.AbstractPageCheck;
 import org.sonar.plugins.html.node.TagNode;
 
 @Rule(key = "TableHeaderHasIdOrScopeCheck")
 public class TableHeaderHasIdOrScopeCheck extends AbstractPageCheck {
 
+  private static final String MESSAGE = "Add either an 'id' or a 'scope' attribute to this <th> tag.";
+
+  private Deque<TableElement> tables = new ArrayDeque<>();
+
   @Override
   public void startElement(TagNode node) {
-    if (isTableHeader(node) && !node.hasProperty("ID") && !node.hasProperty("SCOPE")) {
-      createViolation(node, "Add either an 'id' or a 'scope' attribute to this <" + node.getNodeName() + "> tag.");
+    if (isTableTag(node)) {
+      tables.push(new TableElement());
+    }
+
+    if (isTrTag(node)) {
+      visitTrNode(node);
     }
   }
 
-  private static boolean isTableHeader(TagNode node) {
-    return "TH".equalsIgnoreCase(node.getNodeName());
+  private void visitTrNode(TagNode node) {
+    List<TagNode> row = node.getChildren();
+    if (!tables.isEmpty() && !row.isEmpty()) {
+
+      TableElement currentTable = tables.peek();
+      for (TagNode element : row) {
+        if (isThTag(element)) {
+          currentTable.headers.add(element);
+        }
+      }
+      currentTable.firstCol.add(row.get(0));
+
+      if (currentTable.firstRow.isEmpty()) {
+        currentTable.firstRow = row;
+      }
+
+    } else {
+      // Sometimes rows are defined in separate files, and at this point we don't see them as part of a current table.
+      // In this case we treat them as belonging to "not simple tables".
+      raiseIssueOnTableHeadersWithoutScopeOrId(row);
+    }
+  }
+
+  @Override
+  public void endElement(TagNode node) {
+    if (isTableTag(node) && !tables.isEmpty()) {
+      TableElement currentTable = tables.peek();
+      if (!currentTable.isSimpleTable()) {
+        raiseIssueOnTableHeadersWithoutScopeOrId(currentTable.headers);
+      }
+      tables.pop();
+    }
   }
 
```

Comment:
As a defensive measure, I would override `endDocument()` and put a `tables.clear()` in there. 

As it is the same check instance that gets used for every file, this would avoid cross-file state contamination (<- sounds like a vulnerability :laughing:) when a table does not get closed.
For example, if the first file being scanned is
```html
<table>
    <tr>
        <th>foo</th>
        <th>bar</th>
    </tr>
``` 
and the second is:
```html
<tr>
    <th>foo</th>
    <th>bar</th>
</tr>
</table>
```
This could lead to behavior that might be not so nice to debug in the future. Also memory-wise we would avoid the risk of keeping references to the nodes of a previous file.

Reply:
Indeed, good point, thank you!